src/tests/units/tag_tests.cpp

```diff
@@ -0,0 +1,259 @@
+        checkSum += softAlign[t][s];
+      }
+      REQUIRE(checkSum == Approx(1.0));
+      checkSum = 0;
+    }
+  }
+
+  SECTION("The tag is wrapped in the whole sentence") {
+    // [original]: <span>A republican strategy to counteract the re-election of Obama.</span>
+    // [translated]:<span>Eine republikanische Strategie, um der Wiederwahl Obamas entgegenzuwirken.</span>
+    std::vector<TagNode> tagTree{TagNode({0, 14}, {})};
+
+    std::vector<TagNode> tagTreeTargetExpected{TagNode({0, 17}, {})};
+
+    testCase(softAlign, srcLength, tgtLength, tagTree, tagTreeTargetExpected);
+  }
+
+  SECTION("The tag is wrapped in the middle of the sentence") {
+    // [original]: <span>A republican strategy</span> to counteract the re-election of Obama.
+    // [translated]:<span>Eine republikanische Strategie</span>, um der Wiederwahl Obamas entgegenzuwirken.
+    std::vector<TagNode> tagTree{TagNode({0, 4}, {})};
+    std::vector<TagNode> tagTreeTargetExpected{TagNode({0, 5}, {})};
+
+    testCase(softAlign, srcLength, tgtLength, tagTree, tagTreeTargetExpected);
+  }
+
+  SECTION("The tag is wrapped in the middle of the sentence case 2") {
+    // [original]: A <span>republican strategy</span> to counteract the re-election of Obama.
+    // [translated]:Eine <span>republikanische Strategie</span>, um der Wiederwahl Obamas entgegenzuwirken.
+    std::vector<TagNode> tagTree{TagNode({1, 4}, {})};
+
+    std::vector<TagNode> tagTreeTargetExpected{TagNode({1, 5}, {})};
+
+    testCase(softAlign, srcLength, tgtLength, tagTree, tagTreeTargetExpected);
+  }
+
+  SECTION("The tag is wrapped in one word [one-to-one]") {
+    // [original]: <span>A</span> republican strategy to counteract the re-election of Obama.
+    // [translated]:<span>Eine</span> republikanische Strategie, um der Wiederwahl Obamas entgegenzuwirken.
+    std::vector<TagNode> tagTree{TagNode({0, 1}, {})};
+    std::vector<TagNode> tagTreeTargetExpected{TagNode({0, 1}, {})};
+
+    testCase(softAlign, srcLength, tgtLength, tagTree, tagTreeTargetExpected);
+  }
+
+  //  SECTION("The tag is wrapped in one word [one-to-zero]") {
+  //    // [original]: A republican strategy to counteract the re<span>-</span>election of Obama.
+  //    // [translated]: Eine republikanische Strategie, um der Wiederwahl Obamas entgegenzuwirken.
+  //    std::vector<TagNode> tagTree{
+  //            TagNode({9, 10}, {})
+  //    };
+  //    // should be empty
+  //    std::vector<TagNode> tagTreeTargetExpected{
+  //            TagNode({0, 0}, {})
+  //    };
+  //
+  //    testCase(softAlign, srcLength, tgtLength, tagTree, tagTreeTargetExpected);
+  //  }
+
+  SECTION("Nested tags with one child") {
+    // [original]: <span><span>A republican strategy</span> to counteract the re-election of Obama.</span>
+    // [translated]:<span><span>Eine republikanische Strategie</span>, um der Wiederwahl Obamas entgegenzuwirken.</span>
+    std::vector<TagNode> tagTree{TagNode({0, 14}, {1}), TagNode({0, 4}, {})};
```

Comment:
The way of constructing this should change, this is very unnatural. This should be something like: 

```cpp

class TagTree {
public:
	TagTree(TagNode root){
		vertices_.push_back(root);
	}
	void addChild(TagNode parent, TagNode child){
		vertices.push_back(child);
		parent.children.push_back(vertices.size()-1);
		// I should be able to place asserts here that illegal nesting does not happen.
		// assert (parent.l <= child.l and parent.r >= child.r); 
	} 
	
	const std::vector<TagNode> &vertices(){ return vertices_; }
	
private:
	std::vector<TagNode> vertices_;
};

TagTree tree;
TagNode root = tree.root({0, 14});
TagNode child = tree.addChild(root, {0, 4}}; 
```

The above structure is not ideal (I wouldn't mind pointers for parent and children if there are no complaints of allocations), but externally, something like this provides a more semantically meaningful API to construct the structure from DOM. `TagTree` (even if it's `vector<TagNode>`), `TagNode` needs to be documented heavily.

Reply:
In essence:
1. Any TagTree I construct through the API should be valid.
2. Any nodes I add are valid (children only to existing parents, with range constraints possibly checked). 
3. I should be able to programmatically build this using methods that are natural iterating through DOM elements. Currently I have to do one pass DOM, get the values and then create the `vector<TagNode>` which is not ideal.

Reply:
My current version is only some experimental code which helps to get inside-outside algorithm work. The important part I would like to get from the browser (from Mozilla side) is a tree structure that contains tag information. Because it helps to preserve the tag nesting constraints. For the detailed design of API, let's get inputs from Mozilla side to see what works best.

Reply:
Just two things to confirm. (i) Each validation check could take significant time. It can be as long as O(n) where n is the number of existing nodes in the tree (or O(log n) with an error-prone algorithm and additional space). The total time usage to build the whole TagTree would rise to O(n^2) (or O(n log n) with an error-prone algorithm). I think the user's DOM->tree conversion algorithm should naturally ensure tree validness with zero additional complexity. Is it a good trade to ensure the validness by ourselves at the price of higher complexity or lower maintainability? (ii) I don't mind using pointers to support the physical structure, but in this case, the addChild function would have a pointer-type argument. Is it fine to expose pointers to users? There are several ways to conceal the pointer, but I don't find a particularly good one.

Reply:
> The important part I would like to get from the browser (from Mozilla side) is a tree structure that contains tag information. Because it helps to preserve the tag nesting constraints.

@qianqianzhu I am thinking purely from the perspective of the changes required in the main interface class ([Service](https://github.com/browsermt/bergamot-translator/blob/main/src/translator/service.h#L42)) for this feature to work. I think, providing tag information as a list of `ByteRange` (e.g. `vector<ByteRange>`) should suffice. Right?

Of course, `vector<ByteRange>` can then easily be converted to any representation (e.g. tree structure) internally on which your algorithm can run efficiently. It would be great to keep the [Service](https://github.com/browsermt/bergamot-translator/blob/main/src/translator/service.h#L42) interface as simple as possible + passing a `vector<ByteRange>` instead of a tree structure from the browser would be more efficient.

Reply:
@abhi-agg `vector<ByteRange>` would be OK.  I will do my best to convert this (using the begin and end info from `ByteRange`) into a tree structure.

Reply:
That's great. Thanks. If you get stuck somewhere then please feel free to drop a message 👍

Reply:
@abhi-agg There is one potential problem when using `vector<ByteRange>`: if there are more than one tag around a word(s), the tag nesting order is not clear. E.g., `<p><em>hello world</em></p>`, the corresponding `vector<ByteRange>` is `std::vector<ByteRange> tagVector = {{0,2},{0,2}};`. The nesting order is not clear: we don't know if `tagVector[0]` is wrapped by `tagVector[1]` or `tagVector[1]` is wrapped by `tagVector[0]`. I was assuming the vector of byterange don't hold any tag sequence information. 

To deal with this problem, we could agree on some rule such as **when two byterange are the same in the vector, the first appearing byterange is the outermost tag**. (No expert in HTML. Maybe this nesting order for the same text is not a real problem in HTML.)

Reply:
Can we go slightly stronger?  
The ranges shall always be sorted by:
1. (Primary) Increasing order of begin
2. Increasing order of end
3. Outer nesting

This way a depth first left-first traversal of the tag tree is equivalent to iterating through the vector in order.

Reply:
> The ranges shall always be sorted by:
> 
>     1. (Primary) Increasing order of begin
> 
>     2. Increasing order of end
> 
>     3. Outer nesting
> 
> This way a depth first left-first traversal of the tag tree is equivalent to iterating through the vector in order.

For clarity, what would be the expected `std::vector<ByteRange>` for the following html text:

Original HTML text:         `<div><br><b>abc def<i>ghi jkl</i></b> <u>stu vw</u></div>.`
Text after stripping tags:  `abc defghi jkl stu vw.`
Assuming:
ByteRange of tag `<i>` = [7, 14)
ByteRange of tag `<b>` = [0, 14)
ByteRange of tag `<u>` = [15, 21)
ByteRange of tag `<div>` = [0, 21)
ByteRange of tag `<br>` = [0, )  => please note that the tag `<br>` is an unclosed html tag

Reply:
`<br>` and `<img>` are more of a self-closing HTML tag.  
So I would do:
`<div><br><b>abc def<i>ghi jkl</i></b> <u>stu vw</u></div>`
maps to 
`[0, 21) [0,0) [0,14) [7,14) [15, 21)`
A simpler way of saying things is: the byte ranges should appear in the same order as the open tags in HTML.  

If you have the case `A <span>sentence. Second one.</span>` then after sentence splitting it acts like the tags close at the end of sentence and open at the beginning of the next sentence `A <span>sentence.</span>` and `<span>Second one.</span>`

Reply:
@kpu thanks for your clarification.

@abhi-agg just one additional note about the unclosed/empty html tag: `begin = end` always holds and the tag is placed before the character. Based on your example,

`<br>abc`, the byterange for `<br>` is [0,0)

`a<br>bc`, the byterange for `<br>` is [1,1)

> The byte ranges should appear in the same order as the open tags in HTML.

I will follow this rule to construct TagTree. Since Marian’s alignment works on the token-level rather than character-level (which is passed from the browser), I need to find a way to bridge them.